\documentclass{beamer}

\input{../../slides/packages.tex}
\usepackage{tikz}

\usetheme{Warsaw}

%% Themes
% Outer themes
\useoutertheme{shadow}
% Rounded boxes and shadows
\useinnertheme[shadow=true]{rounded}
% Solid \item symbols
\useinnertheme{circles}

%% Custom colors
\definecolor{rltgreen}{rgb}{0,0.5,0}
\definecolor{pasteur}{RGB}{0,90,154}
\setbeamerfont{block title}{size={}}
\setbeamercolor{structure}{fg=pasteur}
\setbeamercolor{item}{fg=pasteur}


%Color of title
\setbeamertemplate{frametitle}
{    \nointerlineskip
    \begin{beamercolorbox}[sep=0.3cm,ht=1.8em,wd=\paperwidth]{frametitle}
        \vbox{}\vskip-2ex%
        \strut\insertframetitle\strut
        \vskip-0.8ex%
    \end{beamercolorbox}
}
% Hide navigation symbols
\setbeamertemplate{navigation symbols}{}

%% Title block
\setbeamercolor*{title}{use=structure,fg=white,bg=pasteur}

\makeatletter

%% Top infolines
\setbeamertemplate{headline}{%
\leavevmode%
  \hbox{%
    \begin{beamercolorbox}[wd=\paperwidth,ht=2.5ex,dp=1.125ex]{palette quaternary}%
    \insertsectionnavigationhorizontal{\paperwidth}{}{\hskip0pt plus1filll}
    \end{beamercolorbox}%
  }
}


\input{../../slides/snakemake_style.tex}

%% Set up title ---------------------------------------------------------------

\title[Sequana]{Sequana: a set of flexible genomic pipelines for processing 
and reporting NGS analysis}
\author[T. Cokelaer]{Thomas Cokelaer - Dimitri Desvillechabrol}
\institute{Institut Pasteur}
\date{Oct 19th 2017, ENS, Paris (séminaire bioinfo)}


\AtBeginSection[]{
  \begin{frame}
  \vfill
  \centering
  \begin{beamercolorbox}[sep=8pt,center,shadow=true,rounded=true]{title}
    \usebeamerfont{title}\insertsectionhead\par%
  \end{beamercolorbox}
  \vfill
  \end{frame}
}

\begin{document}

%% Title slide ----------------------------------------------------------------

\begin{frame}[plain]
    \titlepage
    \begin{textblock*}{5cm}(4.5cm,0.3cm)
        \includegraphics[scale=0.09]{images/Institut_Pasteur.png}
    \end{textblock*}
\end{frame}

%% Slides ---------------------------------------------------------------------


%\input{../../slides/slide_biomics_2017_jan_1}
%\input{../../slides/slide_motivation_needs_2016_march}
%\input{slide_sequana_motivation}



\begin{frame}
\frametitle{Motivation}
\begin{block}{Jan 2015: provide NGS pipelines to Biomics sequencing platform 
https://research.pasteur.fr/en/team/biomics/ (Institut Pasteur)}
 \begin{itemize}
  \item Genomics: QC + variant calling + de-novo
  \item Transcriptomics: RNA-seq + ChIP-seq 
  \item Metagenomics
  \item Illumina but also Pacbio long reads technologies
 \end{itemize}
 %\includegraphics{images/genetic_strand.png}
 %\includegraphics{images/strand.png}
\end{block} 
\end{frame}



\begin{frame}
\frametitle{How ?}

\begin{columns}
\begin{column}{1.5cm}
\includegraphics[height=0.2\textheight]{images/logo_python.png} 
\end{column}
\begin{column}{9cm}
a glue language, a scientific language
\end{column}
\end{columns}


\rule{\textwidth}{1pt}


\begin{columns}
\begin{column}{1.5cm}
\includegraphics[height=0.2\textheight]{images/logo_snakemake.png}
\end{column}
\begin{column}{9cm}
a pipeline 
framework mixing Python and Makefile \\
{\footnotesize \textcolor{blue}{\textit{Köster, Johannes and Rahmann, Sven. 
Snakemake - A scalable 
bioinformatics workflow engine. Bioinformatics 2012.}}}
\end{column}
\end{columns}

\rule{\textwidth}{1pt}


\begin{columns}
\begin{column}{1.5cm}
\includegraphics[height=0.2\textheight]{images/exe.png}
\end{column}
\begin{column}{9cm}
Dedicated standalone such as genome coverage characterisation or a graphical 
user interface for Snakemake pipelines (Sequanix).
%{\footnotesize \textcolor{blue}{\textit{D. Desvillechabrol, C. 
%Bouchier, S. Kennedy, T. Cokelaer Detection and characterization of low 
%and high genome coverage regions .... BioRxiv 
%https://doi.org/10.1101/092478 }}. Submitted to GigaScience journal }
\end{column}
\end{columns}



\end{frame}



\section{Snakemake as a workflow manager}


\begin{frame}
\centering

$\begin{array}{l}
\includegraphics[height=0.18\textheight]{images/logo_snakemake.png}
\end{array}$
{\Huge =}
Makefile 
{\LARGE + }
$\begin{array}{l}
\includegraphics[height=0.18\textheight]{images/logo_python.png}
\end{array}$


 
\end{frame}



\begin{frame}[fragile]
\frametitle{A minimalist snakemake workflow example (GC content)}
\centering

\begin{columns}
 \begin{column}{7.2cm}
 \hspace{-2cm}
\begin{lstlisting}
# list of FastQ files without extension
files = ["A", "B", ....]

rule all:
    input: expand("{data}.png", data=files), 
        "index.html"

rule create_GC_images:
    input: "{data}.fastq.gz"
    output: "{data}.png"
    run: # CODE TO COMPUTE IMAGES
        
rule html_report:
    input: expand("{data}.png", data=files)
    output: "index.html"
    run: # CODE TO CREATE HTML 
\end{lstlisting} 
\end{column}
\begin{column}{3.9cm}
\hspace{1cm}
 \includegraphics[scale=2.2]{./images/gc.png}
\end{column}
\end{columns}
\end{frame}




\begin{frame}[fragile]
\frametitle{A minimalist snakemake workflow example (GC content)} 
\centering

\begin{columns}
 \begin{column}{7.2cm}
 \hspace{-2cm}
\begin{lstlisting}
# Use an external configuration file
import glob
files = glob.glob(config['gc']['directory'] + "*fastq.gz")
files = [f.replace(".fastq.gz", "") 
         for f in filenames]

rule all:
    input: expand("{data}.png", data=files), 
        "index.html"

rule create_GC_images:
    input: "{data}.fastq.gz"
    output: "{data}.png"
    run: # CODE TO COMPUTE IMAGES
        
rule html_report:
    input: expand("{data}.png", data=files)
    output: "index.html"
    run: # CODE TO CREATE HTML 
\end{lstlisting} 
\end{column}
\begin{column}{3.9cm}
\hspace{1cm}
 \includegraphics[scale=2.2]{./images/gc.png}
\end{column}
\end{columns}
\end{frame}




\begin{frame}[fragile]
\begin{block}{configuration file example in YAML format}
 \begin{lstlisting}
##########################################################
# Input parameters for the fractal analysis
#
# :Parameters: 
#
# - size: output image size formatted as NxM where N and M 
#         are integers
# - depth: a integer (e.g. 200)
# - zoom: a positive value e.g. 0.5
# - N: number of random sets
gc:
    - window: 100
    - directory: /home/user/fastq_files
 \end{lstlisting}
\end{block}
\end{frame}


\begin{frame}[fragile]
\frametitle{Execution}

From a shell:
\rule{\textwidth}{1pt}


\begin{lstlisting}[basicstyle=\ttfamily\large]
snakemake -s gc_minimalist.rules
\end{lstlisting}

\rule{\textwidth}{1pt}

More options if a configuration file is required, or execution is on a cluster, 
or \dots something goes wrong.
\end{frame}

\section{Sequana pipelines (an overview)}


\begin{frame}
\frametitle{Pipeline example: quality control pipeline} 
\centering
\includegraphics[height=0.8\textheight, width=\textwidth]{./images/dag.png}
\end{frame}



\begin{frame}
\frametitle{Pipeline example: variant calling} 
\centering
\includegraphics[height=0.8\textheight, 
width=\textwidth]{./images/variant_calling_dag.png}
\end{frame}



\begin{frame}{Pipeline complexity}

    \includegraphics[width=11cm, height=7cm]{images/number_of_rules.png}

\end{frame}


\begin{frame}[fragile]
 \frametitle{Modularization: Factorise and reuse rules}

 
 Local standard rules
\begin{lstlisting}[basicstyle=\ttfamily\small] 
    include: "path_to_rule_file"
\end{lstlisting}

Sequana rules
\begin{lstlisting}[basicstyle=\ttfamily\small]
    from sequana import snaketools as sm
    include: sm.modules['rulegraph']
\end{lstlisting}

Dynamic rules:
\begin{lstlisting}[basicstyle=\ttfamily\small]
    sm.init("quality_control.rules", globals())
    with open(sequana.modules["fastqc_dynamic"], "r") as dynrule:
        exec(dynrule.read())
        
    manager = sm.PipelineManager("quality_control", 
                                 config)

    include: fastqc_dynamic("example1", manager)
    include: fastqc_dynamic("example2", manager)
\end{lstlisting}
\end{frame}


\begin{frame}{Factorization}
\centering
    \includegraphics[width=8cm]{images/rules_reusing.png}
\end{frame}


\begin{frame}
 
Once upon time there was a pipeline ... 

\vspace{1cm}
\pause

The Snakemake file assembles the rules together but all parameters can be 
externalised in a configuration file. 

 \vspace{1cm}
\pause

One need to edit the configuration file ... without typos 
 
 \vspace{1cm}
\pause

One need to launch the Snakemake command ... without typos
 
\end{frame}


\section{Sequanix}
\input{../../slides/slide_sequana_gui_example}

\begin{frame}
\frametitle{Reference} 
 
\textbf{Sequanix: A Dynamic Graphical Interface for Snakemake Workflows} 

\vspace{.5cm}

\textit{Dimitri Desvillechabrol, Rachel Legendre, Claire Rioualen, Christiane 
Bouchier, Jacques van Helden, Sean Kennedy, Thomas Cokelaer}

\vspace{.5cm}
 
https://www.biorxiv.org/content/early/2017/07/12/162701
 
\end{frame}


 
\section{Sequana coverage}


\begin{frame}
\frametitle{Genome coverage}

\textbf{Definition:} The number of reads mapped to a specific position, $b$, 
within the reference genome. \\
\vspace{1em}
\textbf{Notation:} $ C(b)$ also denoted $C_b$\\
\vspace{1em}
\textbf{Theoretical distribution:} Poisson distribution but in practice over 
dispersed. The poisson parameter is distributed according to a Gamma hence 
leading to a negative binomial (See e.g., Linder et al 2013).
\end{frame}


\begin{frame}
\frametitle{\small Bacteria case (low/high $\mu$ components and del. region)}
\begin{center}
\includegraphics[height=0.9\textheight, 
width=1\textwidth]{images/coverage_bacteria.png}
%Problems (see next figures)
%   Low and high frequency trends prevent fitting of any distribution
%    Some regions are deleted.
%    Others are repeated or over covered by several folds
\end{center}
\end{frame}



\begin{frame}
\frametitle{\small Virus case}
\begin{center}
\includegraphics[height=0.9\textheight, 
width=1\textwidth]{images/coverage_virus.png}
%Problems (see next figures)
%   Low and high frequency trends prevent fitting of any distribution
%    Some regions are deleted.
%    Others are repeated or over covered by several folds
\end{center}
\end{frame}


\begin{frame}
\LARGE
\begin{block}{Question: how to automatically detect and characterise under and 
over covered genomic regions}
\end{block}

\end{frame}


\begin{frame}[fragile]
 \begin{verbatim}
1. The algorithm
   1. Detrending (running median)
   2. Mixture model estimation (Gaussian approximation)
   3. Set a statistics (z-score)
   4. Clustering (double threshold)

\end{verbatim}
\end{frame}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}
\frametitle{1. Detrending}
 
We divide $C_b$ by its moving average (MA), or even better its 
running median (RM) defined as 

\begin{equation}
RM_W(b) = \textrm{median}\left({C(b-V), \dots , C(b+V)}\right) \nonumber
\end{equation}

$W$ is the running window and $V=(W-1)/2$.

\begin{block}{The normalised genome coverage}
\begin{equation}
\widetilde{C}_b = \frac{C_b}{RM_W(b)}  \nonumber
\end{equation}
\end{block}

\pause

\begin{block}{Computational note}
Running median is slow due to the sorting task. \\
\textbf{Solution:} a rolling window + a bisection method to insert 
new element in a sorted list + efficient insert/deletion in a 
list (B-tree) helps:  \textbf{Only a few seconds to scan a 3Mb-length genome.}
\end{block} 
\end{frame}


\begin{frame}
\frametitle{Normalised coverage}
\begin{center}
\includegraphics[height=0.9\textheight, 
    width=1\textwidth]{images/coverage_normalised.png}
\end{center} 
\end{frame}




\begin{frame}
 \frametitle{2. Building a statistics}
 
\textbf{Definition:} the normalised data can be decomposed 
into a \textbf{\textcolor{blue}{central}}
distribution $\widetilde{C}^0$ and a set 
of \textbf{\textcolor{blue}{outliers}} $\widetilde{C}^1$ (above and below 
the central distribution)

\begin{equation} 
\widetilde{C}_b  = \left\{ \widetilde{C}^0_b, \{\widetilde{C}^+_b, 
\widetilde{C}^-_b\}  \right\}
\nonumber
\end{equation}
\vspace{1cm}
\textbf{Hypothesis 1:} Central distribution is predominant.
\begin{equation}
\left|\widetilde{C}^0_b\right| > \left|\widetilde{C}^1_b\right| \nonumber
\end{equation}
\end{frame}



\begin{frame}
\textbf{Hypothesis 2:} The normalised genome coverage follows a 
Gaussian distribution in particular the central distribution

\begin{equation}
PDF(\widetilde{C}_b^0) \sim \mathcal{N}(\mu_0, \sigma_0) \nonumber
\end{equation}

We consider regime with $\delta \gg 1$. However, the algorithm 
works for low values: $\delta \sim 5-10$.
 \end{frame}


\begin{frame}
 
The central distribution $C_0$ will be fitted to a Gaussian 
distribution while the outliers will be fitted to another 
distribution.

\begin{itemize}
\item With an EM algorithm using $k=2$ distributions we can estimate the 
parameters. 
\item The parameters of the \textcolor{blue}{outliers} components are not used. 
\item The average of the \textcolor{blue}{central} distribution is  1 (by 
definition)
\item Note that the average of the \textcolor{blue}{outliers} can be around 1 as
well if the weights of C+ and C- are equivalent
\end{itemize}
\end{frame}


\begin{frame}
\frametitle{}
\begin{center}
\includegraphics[height=0.9\textheight, 
    width=1\textwidth]{images/figure_em.png}
\end{center} 
\end{frame}



\begin{frame}
\frametitle{C. From a constant to adaptative $z$-score}

Assuming that the central distribution is the null hypothesis, 
we can now  assign a \textcolor{blue}{$z$-score in the normalised space}:

\begin{equation}
z(b) = \frac{\widetilde{C}(b)-\widetilde{\mu}_0}{\widetilde{\sigma}_0}
\nonumber
\end{equation}

We can  replace $\widetilde{C}_b$ by its expression  ($C_b / RM_W(b)$) and
express $C_b$ as a function of the running median, the $z(b)$ and the 
parameters 
of the central distribution:

\begin{equation}
C(b) = \left(  \widetilde{\mu}_0  + z(b) \widetilde{\sigma}_0  \right)RM_W(b).
\nonumber
\end{equation}

\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}
\frametitle{C. From a constant to adaptative $z$-score}
\begin{block}{}
We can now set a constant threshold in the normalised space (e.g. $\lambda \pm 
3$) 
and get an adaptative threshold in the original space.
\end{block}

We can 
derive an \textcolor{blue}{adaptative threshold} in the 
original space that is function of the genome position: 

\begin{equation}
\tilde{\delta}^\pm(b) = \left( \widetilde{\mu}_0 \pm \lambda^\pm\times 
\widetilde{\sigma}_0    \right)RM_W(b).
\end{equation}

\end{frame}

\begin{frame}
 \frametitle{constant thresholds in the original space}
\includegraphics[height=0.9\textheight, 
    width=1\textwidth]{images/fig1.png}
 \end{frame}


 \begin{frame}
 \frametitle{constant thresholds in the normalised space}
\includegraphics[height=0.9\textheight, 
    width=1\textwidth]{images/fig2.png}
\end{frame}

 
\begin{frame}
 \frametitle{adaptative thresholds in the original space}
\includegraphics[height=0.9\textheight, 
    width=1\textwidth]{images/fig3.png}
\end{frame}

 
 
\begin{frame}
\frametitle{D. Regions of interest (clustering)}

\begin{block}{}
On a 3Mb genome with low thresholds the number of outliers are high. 
$\lambda=3$ means about 4000 events by pure chance.
\end{block}

\begin{block}{}
We need a strategy to reduce the number of interesting events. This is 
achieved by clustering the data.
\end{block}

\begin{block}{}
Besides, to cluster close-by events further, we proceed with a 
double-thresholds 
approach.
\end{block}
 
\end{frame}


\begin{frame}
 \frametitle{Double thresholds}
\includegraphics[height=0.9\textheight, 
    width=1\textwidth]{images/double_threshold.png}
\end{frame}



\begin{frame}[fragile]
\begin{lstlisting}{python}
sequana_coverage --download-genbank JB409847
sequana_coverage --input JB409847.bed -w 3001 --circular 
                 --genbank JB409847.gbk --show-html
\end{lstlisting}
\includegraphics[height=0.7\textheight, 
    width=1\textwidth]{images/virus.png}
\end{frame}


\begin{frame}[fragile]
 \frametitle{Summary}
\begin{verbatim}
- A robust and fast algo. to detect under/over covered regions.

- The algorithm is made of 3 steps:
  1. Normalisation (running median)
  2. Set a statistics using EM to estimate mixture model
  3. Clustering of events in original space

- Implementation in Sequana as a standalone 
  - HTML reports with ROIs provided as sortable tables
  - Identify repeated regions
  - A genbank can be provided to annotate the report
\end{verbatim}    

 \footnotesize{
Dimitri Desvillechabrol, Christiane Bouchier, Sean Kennedy, Thomas Cokelaer
\textit{Detection and characterization of low and high genome coverage 
regions using an efficient running median and a double threshold approach.}
bioRxiv 092478; doi: http://dx.doi.org/10.1101/092478
}

\end{frame}








 

\section{Sequana: Continuous integration}
\input{../../slides/slide_sequana_ci_2017_oct}

\section{Summary}



\begin{frame}
\frametitle{Sequana in a nutshell}
\vspace{2cm}

$\begin{array}{l}
\includegraphics[height=0.18\textheight]{images/logo_python.png}
\end{array}$
{\Huge +}
$\begin{array}{l}
\includegraphics[height=0.18\textheight]{images/logo_snakemake.png}
\end{array}$
{\Huge +}
$\begin{array}{l}
\includegraphics[height=0.18\textheight]{images/exe.png}
\end{array}$
{\LARGE = Sequana}

\centering
\vspace{2cm}
\Large
http://sequana.readthedocs.io

\end{frame}



\begin{frame}
 \frametitle{Summary}
Sequana is a versatile tool that provides

\begin{enumerate}
 \item A Python library dedicated to NGS analysis (e.g., tools to visualise 
standard NGS formats).
\item A set of snakemake workflows and rules dedicated to NGS 
 \item A GUI to execute them easily with Sequanix
 \item HTML reports
 \item Standalone applications:
 \begin{itemize}
    \item \textbf{sequana\_coverage} ease the extraction of genomic regions of interest and genome coverage information
    \item \textbf{sequana\_taxonomy} get a quick overview of read contents
    \item \dots
 \end{itemize}
\end{enumerate}
\end{frame}



\begin{frame}
\frametitle{You like it ? Please, add a star on our github}
    https://github.com/sequana/sequana/stargazers
    \centering
    \includegraphics[scale=0.28]{images/stargazers.png}\\
\end{frame}


\begin{frame}
\frametitle{You like it ? }
\centering
Join us ! add rules and pipelines !
\end{frame}



    
    



\begin{frame}
 \frametitle{Acknowledgements}

 \begin{itemize}
  \item Dimitri Desvillechabrol (variant calling, denovo, sequana, sequanix)
  \item Rachel (Transcriptomics)
  \item M\'elissa Cardon (pacbio)
  \item Biomics users (Institut Pasteur)
 \end{itemize}
 
\end{frame}


\begin{frame}
\centering
\vspace{2cm}
\LARGE
 Thank you 
 
 \vspace{2cm}
 
 \small
 Slides available on http://github.com/sequana/sequana\_presentations/2017/ENS
\end{frame}




\end{document}
